### PR TITLE
hypeman build

### PR DESCRIPTION
**Pull request by hiroTamada**
> [!NOTE]
> Adds a secure source-to-image build system running in microVMs with API, queuing, storage, and registry integration.
> 
> - New `lib/builds` package: queue, storage, cache keys, metrics, registry JWTs, vsock protocol, file-based secrets, and a guest `builder_agent`
> - API: adds `BuildManager` to `ApiService`, new endpoints (`POST /builds`, `GET /builds`, `GET /builds/{id}`, `DELETE /builds/{id}`, `GET /builds/{id}/logs` via SSE events) and server startup for build services
> - DI/wiring: provides `BuildManager` via `wire` and starts it in `main`
> - Config: build-related settings (`MAX_CONCURRENT_SOURCE_BUILDS`, `BUILDER_IMAGE`, `REGISTRY_URL`, `BUILD_TIMEOUT`, `BUILD_SECRETS_DIR`)
> - Registry/middleware: JWT-scoped repo tokens, Basic/Bearer handling for `/v2/*`, repo/scope validation, and internal VM fallback; tests updated to include full host in image refs
> - OCI: improve unpack/metadata to support Docker v2 → OCI via `go-containerregistry`
> - Tooling: Makefile targets for building the builder image and an E2E build test
> - Tests/docs: extensive unit tests and READMEs for the build system and builder image
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 528cf8dba85f064390fee1d6dc761bef09b335ad. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: lib/builds/images/README.md [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+
+## Building the Generic Builder Image
+
+> **Important**: Hypeman uses `umoci` for OCI image manipulation, which requires images
+> to have **OCI manifest format** (not Docker v2 format). You must use `docker buildx`
+> with the `oci-mediatypes=true` option.
+
+### Prerequisites
+
+1. **Docker Buildx** with a container builder:
```
Comment (hiroTamada) [inline lib/builds/images/README.md]: TODO: we need a script or tooling for creating a new builder image

File: lib/middleware/oapi_auth.go [hunk 2]
```diff
@@ -138,7 +320,7 @@ func JwtAuth(jwtSecret string) func(http.Handler) http.Handler {
 				return
 			}
 
-			// Parse and validate JWT
+			// Parse and validate as regular user JWT
```
[the remark this answers is not in the document]
Comment (hiroTamada) [reply lib/middleware/oapi_auth.go:338]: well builder vm does not have access to `/home/builder/.docker/config.json`. But I guess you are right

File: openapi.yaml [hunk 3]
```diff
@@ -683,6 +683,133 @@ components:
...
+          type: string
+          description: SHA256 hash of source tarball
+        lockfile_hashes:
+          type: object
+          additionalProperties:
+            type: string
+          description: Map of lockfile names to SHA256 hashes
+        toolchain_version:
+          type: string
+          description: Runtime version (e.g., "node v20.10.0")
```
Comment (rgarcia) [inline openapi.yaml]: on first read not sure what this means for a generic build

File: openapi.yaml [hunk 4]
```diff
@@ -683,6 +683,133 @@ components:
...
+          format: date-time
+          description: Build completion timestamp
+    
+    CreateBuildRequest:
+      type: object
+      properties:
+        runtime:
+          type: string
+          description: "(Deprecated) Build runtime hint. No longer required - the generic builder accepts any Dockerfile."
+          example: generic
```
Comment (rgarcia) [inline openapi.yaml]: can remove?

File: openapi.yaml [hunk 5]
```diff
@@ -683,6 +683,133 @@ components:
...
+        cache_scope:
+          type: string
+          description: Tenant-specific cache key prefix for isolation
+        build_args:
+          type: object
+          additionalProperties:
+            type: string
+          description: Build arguments to pass to Dockerfile
+        build_policy:
+          $ref: "#/components/schemas/BuildPolicy"
```
Comment (rgarcia) [inline openapi.yaml]: on first read surprised to not see the tarball in this request... in fact i don't see CreateBuildRequest used anywhere?

File: openapi.yaml [hunk 6]
```diff
@@ -1936,4 +2063,200 @@ paths:
...
+                source:
+                  type: string
+                  format: binary
+                  description: Source tarball (tar.gz) containing application code and optionally a Dockerfile
+                dockerfile:
+                  type: string
+                  description: Dockerfile content. Required if not included in the source tarball.
+                runtime:
+                  type: string
+                  description: "(Deprecated) Build runtime hint. No longer required."
```
Comment (rgarcia) [inline openapi.yaml]: can remove?

File: openapi.yaml [hunk 7]
```diff
@@ -1936,4 +2063,200 @@ paths:
...
+            type: boolean
+            default: false
+          description: Continue streaming new lines after initial output
+      responses:
+        200:
+          description: Log stream (SSE)
+          content:
+            text/event-stream:
+              schema:
+                type: string
```
Comment (rgarcia) [inline openapi.yaml]: it might be nice to have an object here or a union of different objects, e.g. make the endpoint /build/{id}/events where LogEvent is a json object with a timetstamp and content. A BuildStatusEvent that signals a change in status (building->pushing->success). There's also some benefit to having all SSE endpoints send a heartbeat event since cloudflare and the like will kill idle connections after a period of inactivity e.g. 100s. We do this in the main Kernel API to avoid this issue

File: lib/builds/images/generic/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,59 @@
+# Generic Builder Image
+# Contains rootless BuildKit + builder agent - runtime agnostic
+# Users provide their own Dockerfile which specifies the runtime (node, python, etc.)
```
Comment (rgarcia) [inline lib/builds/images/generic/Dockerfile]: might remove meta references like this to previous non-generic setup

File: lib/builds/builder_agent/main.go [hunk 9]
```diff
@@ -0,0 +1,520 @@
...
+
+const (
+	configPath = "/config/build.json"
+	vsockPort  = 5001 // Build agent port (different from exec agent)
+)
+
+// BuildConfig matches the BuildConfig type from lib/builds/types.go
+type BuildConfig struct {
+	JobID           string            `json:"job_id"`
+	Runtime         string            `json:"runtime"`
```
Comment (rgarcia) [inline lib/builds/builder_agent/main.go]: can remove?

File: lib/builds/builder_agent/main.go [hunk 10]
```diff
@@ -0,0 +1,520 @@
...
+	cmd := exec.Command("buildctl", "--version")
+	out, _ := cmd.Output()
+	return strings.TrimSpace(string(out))
+}
+
+func getToolchainVersion() string {
+	// Generic builder doesn't have runtime-specific toolchains
+	// The actual runtime version is determined by the user's Dockerfile
+	return "generic"
+}
```
Comment (rgarcia) [inline lib/builds/builder_agent/main.go]: can remove this and things that use it?

File: lib/builds/builder_agent/main.go [hunk 11]
```diff
@@ -0,0 +1,520 @@
...
+	buildResult     *BuildResult
+	buildResultLock sync.Mutex
+	buildDone       = make(chan struct{})
+)
+
+func main() {
+	log.Println("=== Builder Agent Starting ===")
+
+	// Start vsock listener first (so host can connect as soon as VM is ready)
+	listener, err := startVsockListener()
```
Comment (rgarcia) [inline lib/builds/builder_agent/main.go:111]: does the guest-agent also run on builder VMs? I think it'd be useful to have it so that you can exec into a failed build or do other debugging

File: lib/builds/builder_agent/main.go [hunk 12]
```diff
@@ -0,0 +1,520 @@
...
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, time.Duration(config.TimeoutSeconds)*time.Second)
+		defer cancel()
+	}
+
+	// Note: Secret fetching would need the host connection
+	// For now, we skip secrets if they require host communication
+	// TODO: Implement bidirectional secret fetching
+	if len(config.Secrets) > 0 {
+		log.Printf("Warning: Secrets requested but vsock secret fetching not yet implemented in new model")
```
Comment (rgarcia) [inline lib/builds/builder_agent/main.go]: probably worth it to implement

File: lib/builds/PLAN.md [hunk 13]
```diff
@@ -0,0 +1,182 @@
+# Build System Roadmap
+
+## Current State (v0.1)
+
+- ✅ Source-to-image builds in isolated microVMs
+- ✅ BuildKit-based builds with daemonless execution
+- ✅ Tenant-isolated registry caching
+- ✅ Node.js 20 and Python 3.12 runtimes
```
Comment (rgarcia) [inline lib/builds/PLAN.md]: ❌

File: scripts/e2e-build-test.sh [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+        echo ""
+        log "=== Build Logs ==="
+        get_logs "$TOKEN" "$BUILD_ID"
+        echo ""
+        error "=== E2E Test FAILED ==="
+        
+        # Cleanup
+        rm -f "$SOURCE"
+        exit 1
+    fi
```
Comment (rgarcia) [inline scripts/e2e-build-test.sh:479]: maybe also attempt to run a vm with the built image?

File: lib/builds/images/README.md [hunk 15]
```diff
@@ -0,0 +1,260 @@
...
+```bash
+# Build locally
+docker build \
+  -t hypeman/builder:local \
+  -f lib/builds/images/generic/Dockerfile \
+  .
+
+# Run locally to test
+docker run --rm hypeman/builder:local --help
+```
```
Comment (rgarcia) [inline lib/builds/images/README.md:117]: curious how you think hypeman deployment process should load this builder image... should it be part of the deploy process or should hypeman itself perform this build on startup?

File: lib/builds/vsock_handler.go [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+	// GetSecrets returns the values for the given secret IDs
+	GetSecrets(ctx context.Context, secretIDs []string) (map[string]string, error)
+}
+
+// NoOpSecretProvider returns empty secrets (for builds without secrets)
+type NoOpSecretProvider struct{}
+
+func (p *NoOpSecretProvider) GetSecrets(ctx context.Context, secretIDs []string) (map[string]string, error) {
+	return make(map[string]string), nil
+}
```
Comment (rgarcia) [inline lib/builds/vsock_handler.go:45]: Curious how you're thinking about modeling / implementing build secrets...

File: lib/builds/types.go [hunk 17]
```diff
@@ -0,0 +1,208 @@
...
+	StatusCancelled = "cancelled"
+)
+
+// Runtime constants (deprecated - kept for backward compatibility)
+// The generic builder system no longer requires runtime selection.
+// Users provide their own Dockerfile which specifies the runtime.
+const (
+	RuntimeNodeJS20  = "nodejs20"  // Deprecated
+	RuntimePython312 = "python312" // Deprecated
+)
```
Comment (rgarcia) [inline lib/builds/types.go]: ❌

File: cmd/api/api/builds.go [hunk 18]
```diff
@@ -0,0 +1,320 @@
...
+		if len(b.Provenance.LockfileHashes) > 0 {
+			oapiBuild.Provenance.LockfileHashes = &b.Provenance.LockfileHashes
+		}
+	}
+
+	return oapiBuild
+}
+
+// deref safely dereferences a pointer, returning empty string if nil
+func deref(s *string) string {
```
Comment (rgarcia) [inline cmd/api/api/builds.go]: appears unused

File: scripts/e2e-build-test.sh (hunk 14 above)
Comment (rgarcia) [reply scripts/e2e-build-test.sh:479]: and/or make regular Go tests for the build manager?

**Review by rgarcia — APPROVED** [review]

File: lib/builds/images/README.md (hunk 15 above)
Comment (hiroTamada) [reply lib/builds/images/README.md:117]: I think we can use the same process as kernel-images, and runtime images where you manually publish images to public docker hub.

File: lib/builds/vsock_handler.go (hunk 16 above)
Comment (hiroTamada) [reply lib/builds/vsock_handler.go:45]: I actually implemented this

File: openapi.yaml [hunk 19]
```diff
@@ -683,6 +683,121 @@ components:
...
+          additionalProperties:
+            type: string
+          description: Map of lockfile names to SHA256 hashes
+        buildkit_version:
+          type: string
+          description: BuildKit version used
+        timestamp:
+          type: string
+          format: date-time
+          description: Build completion timestamp
```
Comment (rgarcia) [inline openapi.yaml:732]: nit but maybe `completed_at` would be a better name? Actually maybe this is just redundant with `Build.completed_at` and should be removed? Also I was somewhat expecting the provenance object to contain the started-at time ("provenance" chat tells me, comes from the French provenir, meaning "to come from" or "to originate"). But then it also seems redundant with the timestamp for started at in the main build object.

File: Makefile [hunk 20]
```diff
@@ -206,6 +206,17 @@ test: ensure-ch-binaries ensure-caddy-binaries build-embedded
...
 
+# Build the generic builder image for builds
+build-builder:
+	docker build -t hypeman/builder:latest -f lib/builds/images/generic/Dockerfile .
+
+# Alias for backwards compatibility
+build-builders: build-builder
+
+# Run E2E build system test (requires server running: make dev)
+e2e-build-test:
```
Comment (rgarcia) [inline Makefile:217]: The `build-builders` alias seems premature if there's only one builder image—consider removing it unless there's a concrete plan for multiple builders soon.

File: cmd/api/config/config.go [hunk 21]
```diff
@@ -103,6 +103,13 @@ type Config struct {
 	// Cloudflare configuration (if AcmeDnsProvider=cloudflare)
 	CloudflareApiToken string // Cloudflare API token
 
+	// Build system configuration
+	MaxConcurrentSourceBuilds int    // Max concurrent source-to-image builds
+	BuilderImage              string // OCI image for builder VMs
+	RegistryURL               string // URL of registry for built images
+	BuildTimeout              int    // Default build timeout in seconds
```
Comment (rgarcia) [inline cmd/api/config/config.go:110]: This has to be the hypeman registry URL, right? Since it relies on custom auth specific to hypeman, maybe it should not be configurable.

File: cmd/api/api/builds.go [hunk 22]
```diff
@@ -0,0 +1,313 @@
...
+			return oapi.CreateBuild400JSONResponse{
+				Code:    "invalid_request",
+				Message: "failed to parse multipart form",
+			}, nil
+		}
+
+		switch part.FormName() {
+		case "source":
+			sourceData, err = io.ReadAll(part)
+			if err != nil {
```
Comment (rgarcia) [inline cmd/api/api/builds.go:63]: The multipart parsing loop has no size limit on `sourceData`—could be a DoS vector if someone uploads a huge file. Should there be a max size check?

File: cmd/api/api/builds.go [hunk 23]
```diff
@@ -0,0 +1,313 @@
...
+
+	eventChan, err := s.BuildManager.StreamBuildEvents(ctx, request.Id, follow)
+	if err != nil {
+		if errors.Is(err, builds.ErrNotFound) {
+			return oapi.GetBuildEvents404JSONResponse{
+				Code:    "not_found",
+				Message: "build not found",
+			}, nil
+		}
+		log.ErrorContext(ctx, "failed to stream build events", "error", err, "id", request.Id)
```
Comment (rgarcia) [inline cmd/api/api/builds.go:246]: If `json.Marshal(event)` fails here, it silently continues. Might be worth logging this error.

File: cmd/api/api/builds.go [hunk 24]
```diff
@@ -0,0 +1,313 @@
...
+		}, nil
+	}
+
+	return oapi.GetBuild200JSONResponse(buildToOAPI(build)), nil
+}
+
+// CancelBuild cancels a build
+func (s *ApiService) CancelBuild(ctx context.Context, request oapi.CancelBuildRequestObject) (oapi.CancelBuildResponseObject, error) {
+	log := logger.FromContext(ctx)
+
```
Comment (rgarcia) [inline cmd/api/api/builds.go:200]: This returns 409 for `ErrBuildInProgress` with message "already in progress"—but usually you'd cancel a build *because* it's in progress. Is this the right semantic? Should this case allow cancellation, or is the error name/message misleading?

File: lib/builds/types.go [hunk 25]
```diff
@@ -0,0 +1,210 @@
...
+	CPUs int `json:"cpus,omitempty"`
+
+	// NetworkMode controls network access during build
+	// "isolated" = no network, "egress" = outbound allowed
+	NetworkMode string `json:"network_mode,omitempty"`
+
+	// AllowedDomains restricts egress to specific domains (only when NetworkMode="egress")
+	AllowedDomains []string `json:"allowed_domains,omitempty"`
+}
+
```
Comment (rgarcia) [inline lib/builds/types.go:75]: `AllowedDomains` is defined here but doesn't appear in `BuildConfig` that gets passed to the VM. Is domain filtering implemented, or is this a placeholder for future work?

File: lib/builds/errors.go [hunk 26]
```diff
@@ -0,0 +1,35 @@
...
+	ErrInvalidSource = errors.New("invalid source")
+
+	// ErrSourceHashMismatch is returned when the source hash doesn't match
+	ErrSourceHashMismatch = errors.New("source hash mismatch")
+
+	// ErrBuilderNotReady is returned when the builder image is not available
+	ErrBuilderNotReady = errors.New("builder image not ready")
+
+	// ErrBuildInProgress is returned when trying to cancel a build that's already complete
+	ErrBuildInProgress = errors.New("build in progress")
```
Comment (rgarcia) [inline lib/builds/errors.go:34]: The comment says "returned when trying to cancel a build that's already complete" but the error name is `ErrBuildInProgress`—these are contradictory. Should this be `ErrBuildAlreadyComplete` or similar?

File: lib/builds/queue.go [hunk 27]
```diff
@@ -0,0 +1,171 @@
...
+		startFn()
+	}
+
+	build := QueuedBuild{
+		BuildID: buildID,
+		Request: req,
+		StartFn: wrappedFn,
+	}
+
+	// Start immediately if under concurrency limit
```
Comment (rgarcia) [inline lib/builds/queue.go:73]: `Enqueue` launches goroutines with `go wrappedFn()`. If the queue is stopped/shutdown, there's no way to cancel these goroutines. Should `startFn` accept a context for cancellation?

File: lib/builds/queue.go [hunk 28]
```diff
@@ -0,0 +1,171 @@
...
+// Design notes (see plan for full context):
+// - Queue state is in-memory (lost on restart)
+// - Build metadata is persisted to disk
+// - On startup, pending builds are recovered via listPendingBuilds()
+//
+// Future migration path if needed:
+// - Add BuildQueue interface with Enqueue/Dequeue/Ack/Nack
+// - Implement adapters: memoryQueue, redisQueue, natsQueue
+// - Use BUILD_QUEUE_BACKEND env var to select implementation
+type BuildQueue struct {
```
Comment (rgarcia) [inline lib/builds/queue.go:24]: The design notes mention "pending builds are recovered via listPendingBuilds()" but this queue is purely in-memory. Recovery would need to re-enqueue builds, which could cause issues if a `startFn` had already partially executed before the restart. Is there handling for this edge case?

File: lib/builds/storage.go [hunk 29]
```diff
@@ -0,0 +1,243 @@
...
+
+	var metas []*buildMetadata
+	for _, entry := range entries {
+		if !entry.IsDir() {
+			continue
+		}
+
+		meta, err := readMetadata(p, entry.Name())
+		if err != nil {
+			continue // Skip invalid entries
```
Comment (rgarcia) [inline lib/builds/storage.go:111]: `listAllBuilds` silently skips invalid entries here. Might be worth logging these errors for debugging—a corrupted metadata file would be silently ignored.

File: lib/builds/cache.go [hunk 30]
```diff
@@ -0,0 +1,179 @@
...
+
+	return hex.EncodeToString(h.Sum(nil))
+}
+
+// sortedKeys returns the keys of a map in sorted order
+func sortedKeys(m map[string]string) []string {
+	keys := make([]string, 0, len(m))
+	for k := range m {
+		keys = append(keys, k)
+	}
```
Comment (rgarcia) [inline lib/builds/cache.go:154]: This uses bubble sort with the comment "for small maps"—but `sort.Strings` from stdlib would be clearer and no slower for small slices.

File: lib/builds/file_secret_provider.go [hunk 31]
```diff
@@ -0,0 +1,66 @@
...
+// GetSecrets returns the values for the given secret IDs by reading files from the secrets directory.
+// Missing secrets are silently skipped (not an error).
+// Returns an error only if a secret file exists but cannot be read.
+func (p *FileSecretProvider) GetSecrets(ctx context.Context, secretIDs []string) (map[string]string, error) {
+	result := make(map[string]string)
+
+	for _, id := range secretIDs {
+		// Validate secret ID to prevent path traversal
+		if strings.Contains(id, "/") || strings.Contains(id, "\\") || id == ".." || id == "." {
+			continue // Skip invalid IDs
```
Comment (rgarcia) [inline lib/builds/file_secret_provider.go:35]: The path traversal check skips invalid IDs silently. This could mask user errors—if someone requests a secret `my/token` they'd get no feedback that the ID is invalid. Consider returning an error or at least logging.

File: lib/builds/manager.go [hunk 32]
```diff
@@ -0,0 +1,1078 @@
...
+		NetworkMode:     policy.NetworkMode,
+	}
+	if err := writeBuildConfig(m.paths, id, buildConfig); err != nil {
+		deleteBuild(m.paths, id)
+		return nil, fmt.Errorf("write build config: %w", err)
+	}
+
+	// Enqueue the build
+	queuePos := m.queue.Enqueue(id, req, func() {
+		m.runBuild(context.Background(), id, req, policy)
```
Comment (rgarcia) [inline lib/builds/manager.go:227]: This uses `context.Background()` instead of the parent context, so if the API request is cancelled the build still runs. If this is intentional, consider using `context.WithoutCancel(ctx)` instead so that context values (like OpenTelemetry trace context) are still propagated.

File: lib/builds/manager.go [hunk 33]
```diff
@@ -0,0 +1,1078 @@
...
+func (m *manager) waitForResult(ctx context.Context, inst *instances.Instance) (*BuildResult, error) {
+	// Wait a bit for the VM to start and the builder agent to listen on vsock
+	time.Sleep(3 * time.Second)
+
+	// Try to connect to the builder agent with retries
+	var conn net.Conn
+	var err error
+
+	for attempt := 0; attempt < 30; attempt++ {
+		select {
```
Comment (rgarcia) [inline lib/builds/manager.go:423]: This will be in the hot path so the 3 second sleep pains me deeply :) Is there some way to avoid this or make it more precise / less fuzzy?

File: lib/builds/manager.go [hunk 34]
```diff
@@ -0,0 +1,1078 @@
...
+	event := BuildEvent{
+		Type:      EventTypeStatus,
+		Timestamp: time.Now(),
+		Status:    status,
+	}
+
+	for _, ch := range m.statusSubscribers[buildID] {
+		// Non-blocking send - drop if channel is full
+		select {
+		case ch <- event:
```
Comment (rgarcia) [inline lib/builds/manager.go:683]: When cancelling a building/pushing job, this deletes the instance but doesn't wait for confirmation. The deferred cleanup in `executeBuild` might race with this—is that okay?

File: lib/builds/builder_agent/main.go [hunk 35]
```diff
@@ -0,0 +1,677 @@
...
+
+	"github.com/mdlayher/vsock"
+)
+
+const (
+	configPath = "/config/build.json"
+	vsockPort  = 5001 // Build agent port (different from exec agent)
+)
+
+// BuildConfig matches the BuildConfig type from lib/builds/types.go
```
Comment (rgarcia) [inline lib/builds/builder_agent/main.go:38]: Maybe add a comment here noting that these types must be kept in sync with `lib/builds/types.go` and are only duplicated since this is a separate binary.

File: lib/builds/builder_agent/main.go [hunk 36]
```diff
@@ -0,0 +1,677 @@
...
+		case <-time.After(30 * time.Second):
+			log.Printf("Warning: Timeout waiting for secrets, proceeding anyway")
+			// Signal secrets ready to avoid blocking other goroutines
+			secretsOnce.Do(func() {
+				close(secretsReady)
+			})
+		case <-ctx.Done():
+			setResult(BuildResult{
+				Success:    false,
+				Error:      "build timeout while waiting for secrets",
```
Comment (rgarcia) [inline lib/builds/builder_agent/main.go:401]: I'd think secrets timeout would be a hard fail? If secrets are required for the build, proceeding without them will just cause a more confusing error later.

File: lib/middleware/oapi_auth.go [hunk 37]
```diff
@@ -15,6 +17,18 @@ type contextKey string
 
 const userIDKey contextKey = "user_id"
 
+// registryPathPattern matches /v2/{repository}/... paths
```
Comment (rgarcia) [inline lib/middleware/oapi_auth.go:20]: Maybe add a comment that this type needs to stay in sync with `lib/builds/registry_token.go` and `lib/builds/builder_agent/main.go`.

File: lib/middleware/oapi_auth.go [hunk 38]
```diff
@@ -124,6 +276,51 @@ func JwtAuth(jwtSecret string) func(http.Handler) http.Handler {
...
+								"build_id", registryClaims.BuildID,
+								"repos", registryClaims.Repositories,
+								"scope", registryClaims.Scope)
+							ctx := context.WithValue(r.Context(), userIDKey, "builder-"+registryClaims.BuildID)
+							next.ServeHTTP(w, r.WithContext(ctx))
+							return
+						}
+						log.DebugContext(r.Context(), "registry token validation failed", "error", err)
+					} else {
+						log.DebugContext(r.Context(), "failed to extract token", "error", err)
```
Comment (rgarcia) [inline lib/middleware/oapi_auth.go:302]: This IP fallback is marked as deprecated—is there a plan/timeline to remove it?

**Review by rgarcia — APPROVED** [review]
Great work on the build system! This is a well-designed, comprehensive feature that adds source-to-image builds in isolated microVMs with proper security boundaries.

**Highlights:**
- Clean architecture with BuildKit running rootless inside ephemeral VMs
- Good tenant isolation via scoped registry tokens and cache keys
- Solid test coverage across core components
- Thorough documentation in the README

I've left some inline comments—mostly minor suggestions and a few questions about edge cases. Nothing blocking.

🚀 Ship it!
